src/Stratis.Bitcoin.Features.SmartContracts/Wallet/SmartContractTransactionService.cs

```diff
@@ -199,37 +200,122 @@ public BuildContractTransactionResult BuildTx(BuildContractTransactionRequest re
+        public static string[] ParseArray(string parameter, ref int position)
         {
-            const int signatureLength = 65;
-            const int minHeaderByte = 27;
-            const int maxHeaderByte = 34;
-
-            if (parameters == null)
-                return null;
+            var elements = new List<string>();
+            int elementStart = position;
+            bool quoted = false;
 
-            // Replace SIG# with any included signatures.
-            string encodedSigs = null;
-            for (int i = 0; i < parameters.Length; i++)
+            for (; position < parameter.Length; position++)
             {
-                if (parameters[i].ToUpper() == "SIG#")
+                if (parameter[position] == '\'')
                 {
-                    if (encodedSigs == null)
-                    {
-                        var sigs = (signatures ?? new string[0]).Select(s => Convert.FromBase64String(s)).ToArray();
-                        if (sigs.Any(s => s.Length != signatureLength || s[0] < minHeaderByte || s[0] > maxHeaderByte))
-                            throw new Exception("Invalid signature(s).");
+                    quoted = !quoted;
+                }
+                else if (!quoted && (parameter[position] == ',' || parameter[position] == ']'))
+                {
+                    string placeHolder = "\x0";
+
+                    // Extract the element.
+                    // Replace double-single-quotes with a placeholder. Remove single-quotes. Replace the placeholder with a single-quote.
+                    string element = parameter.Substring(elementStart, position - elementStart).Trim().Replace("''", placeHolder).Replace("'", "").Replace(placeHolder, "'");
+                    elements.Add(element);
+                    if (parameter[position] == ']')
+                        return elements.ToArray();
+                    elementStart = position + 1;
+                }
+            }
 
-                        var sigbuf = new byte[sigs.Length * signatureLength];
-                        for (int j = 0; j < sigs.Length; j++)
-                            Array.Copy(sigs[j], 0, sigbuf, j * signatureLength, signatureLength);
+            return null;
+        }
 
-                        encodedSigs = $"{(int)MethodParameterDataType.ByteArray}#{Encoders.Hex.EncodeData(sigbuf)}";
-                    }
+        /// <summary>
+        /// If the parameter contains an array it is converted to a byte array.
+        /// </summary>
+        /// <param name="cpSerializer">Contract primitive serializer.</param>
+        /// <param name="mpSerializer">Method parameter string serializer.</param>
+        /// <param name="parameter">The paramter to convert (if required).</param>
+        /// <returns>The converted (or unconverted) parameter.</returns>
+        public static string ConvertParameter(ContractPrimitiveSerializer cpSerializer, MethodParameterStringSerializer mpSerializer, string parameter)
```

Comment:
```
        /// <summary>
        /// If the parameter contains an array it is converted to a byte array.
        /// </summary>
        /// <param name="cpSerializer">Contract primitive serializer.</param>
        /// <param name="mpSerializer">Method parameter string serializer.</param>
        /// <param name="parameter">The paramter to convert (if required).</param>
        /// <returns>The converted (or unconverted) parameter.</returns>
        public static string ConvertParameter(ContractPrimitiveSerializer cpSerializer, MethodParameterStringSerializer mpSerializer, string parameter)
        {
            var typeDigits = parameter.TakeWhile(c => char.IsDigit(c));

            var prefixNumbers = string.Join(string.Empty, typeDigits);

            // If this parameter is not an array then ignore it.
            if (parameter.ElementAtOrDefault(prefixNumbers.Length) != '[')
                return parameter;

            MethodParameterDataType dataType;

            if (prefixNumbers.Length == 0) // default is string if data type not supplied 
            {
                dataType = MethodParameterDataType.String;
            }
            else if (!Enum.TryParse(prefixNumbers, out dataType))
            {
                return parameter;
            }

            // Parse the array.
            // The 'position' should now be set to the first character following '['.
            var elements = ParseArray(parameter, prefixNumbers.Length + 1);

            object[] arrayElements = mpSerializer.Deserialize(elements.Select(e => $"{(int)dataType}#{e}").ToArray());

            var serializedArray = cpSerializer.Serialize(arrayElements);

            var hex = BitConverter.ToString(serializedArray).Replace("-", "");

            return $"{(int)MethodParameterDataType.ByteArray}#{hex}";
        }
```

Reply:
Thanks @YakupIpek . I've taken the nice bits from this code.